pkg/internalapis/csinodetopology/csinodetopology.go

```diff
@@ -16,9 +16,240 @@ limitations under the License.
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, info.NodeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: info.NodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       nodeObj.Name,
+					UID:        nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: info.NodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = volTopology.k8sClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create CSINodeTopology CR. Error: %+v", err)
+		}
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", info.NodeName)
+	}
+
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
+	watchCSINodeTopology, err := volTopology.csiNodeTopologyWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", info.NodeName).String(),
+		TimeoutSeconds: &timeoutSeconds,
+		Watch:          true,
+	})
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to watch on CSINodeTopology instance with name %q. Error: %+v", info.NodeName, err)
+	}
+	defer watchCSINodeTopology.Stop()
+
+	// Check if status gets updated in the instance within the given timeout seconds.
+	for event := range watchCSINodeTopology.ResultChan() {
+		csiNodeTopologyInstance, ok := event.Object.(*csinodetopologyv1alpha1.CSINodeTopology)
+		if !ok {
+			log.Warnf("Received unidentified object - %+v", event.Object)
+			continue
+		}
+		if csiNodeTopologyInstance.Name != info.NodeName {
+			continue
+		}
+		switch csiNodeTopologyInstance.Status.Status {
+		case csinodetopologyv1alpha1.CSINodeTopologySuccess:
+			accessibleTopology := make(map[string]string)
+			for _, label := range csiNodeTopologyInstance.Status.TopologyLabels {
+				accessibleTopology[label.Key] = label.Value
+			}
+			return accessibleTopology, nil
+		case csinodetopologyv1alpha1.CSINodeTopologyError:
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to retrieve topology information for Node: %q. Error: %q", info.NodeName,
+				csiNodeTopologyInstance.Status.ErrorMessage)
+		}
+	}
+	return nil, logger.LogNewErrorCodef(log, codes.Internal,
+		"timed out while waiting for topology labels to be updated in %q CSINodeTopology instance.",
+		info.NodeName)
+}
+
+// getNodeObject fetches the node instance given the nodeName
+func getNodeObject(ctx context.Context, nodeName string) (*v1.Node, error) {
+	log := logger.GetLogger(ctx)
+
+	// Create the kubernetes client.
+	k8sClient, err := k8s.NewClient(ctx)
```

Comment:
How about storing this `k8sclient` too as part of the `volumeTopology` struct, like the client you have for `CSINodeTopology` instances. Then we dont need to keep calling `NewClient` here

Reply:
Cool, will make the change.

Reply:
Done